
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
        <meta name="author" content="Chunyi Wang">
      
      
      
        <link rel="prev" href="..">
      
      
        <link rel="next" href="../Formula/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.8">
    
    
      
        <title>Debug and refinement - Run Monopogen</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.f2e4d321.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#debug-and-refinement" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Run Monopogen" class="md-header__button md-logo" aria-label="Run Monopogen" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Run Monopogen
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Debug and refinement
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="teal" data-md-color-accent="lime"  aria-label="Switch to Light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to Light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/CyrusWang57/run_monopogen_doc" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Homepage

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
    
  
  Debug and refinement

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../Formula/" class="md-tabs__link">
        
  
    
  
  Formula

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../Parameters/" class="md-tabs__link">
        
  
    
  
  Parameters

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../Results/" class="md-tabs__link">
        
  
    
  
  Results

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Run Monopogen" class="md-nav__button md-logo" aria-label="Run Monopogen" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    Run Monopogen
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/CyrusWang57/run_monopogen_doc" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Homepage
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Debug and refinement
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Debug and refinement
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#dependencies" class="md-nav__link">
    <span class="md-ellipsis">
      Dependencies
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Dependencies">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#r-environment" class="md-nav__link">
    <span class="md-ellipsis">
      R environment
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-tools" class="md-nav__link">
    <span class="md-ellipsis">
      Other tools
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sex-chromosomes-is-not-considered-by-default" class="md-nav__link">
    <span class="md-ellipsis">
      Sex chromosomes is not considered by default
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Sex chromosomes is not considered by default">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-way-to-specify-imputation-panel" class="md-nav__link">
    <span class="md-ellipsis">
      The way to specify imputation panel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add-x-chromosome-in-multiple-loops" class="md-nav__link">
    <span class="md-ellipsis">
      Add X chromosome in multiple loops
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#modify-source-file" class="md-nav__link">
    <span class="md-ellipsis">
      Modify source file
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#rewrite-code-spliting-bam-by-cell-barcodes" class="md-nav__link">
    <span class="md-ellipsis">
      Rewrite code spliting bam by cell barcodes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#other-bugs" class="md-nav__link">
    <span class="md-ellipsis">
      Other bugs
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Other bugs">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#havent-generate-index-of-vcf-file" class="md-nav__link">
    <span class="md-ellipsis">
      Haven't generate index of vcf file.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#separator-is-not-specified-in-read_csv" class="md-nav__link">
    <span class="md-ellipsis">
      Separator is not specified in read_csv.
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#indent-error" class="md-nav__link">
    <span class="md-ellipsis">
      Indent error
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#release-the-limit-on-the-number-of-files-that-can-be-opened-simultaneously" class="md-nav__link">
    <span class="md-ellipsis">
      Release the limit on the number of files that can be opened simultaneously
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#issues-in-ldrefinementr" class="md-nav__link">
    <span class="md-ellipsis">
      Issues in LDrefinement.R
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Issues in LDrefinement.R">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#issues-in-define-snv-pairs" class="md-nav__link">
    <span class="md-ellipsis">
      Issues in define SNV pairs
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#issue-in-the-funtion-weighthedp" class="md-nav__link">
    <span class="md-ellipsis">
      Issue in the funtion weighthedP
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#issue-in-adjusting-ld-refinement-score-based-on-their-phasing-information" class="md-nav__link">
    <span class="md-ellipsis">
      Issue in adjusting LD refinement score based on their phasing information
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#paint-germline-ld-refinement-score-plot-of-all-chromosomes-in-a-picture" class="md-nav__link">
    <span class="md-ellipsis">
      Paint germline LD refinement score plot of all chromosomes in a picture
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Formula/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Formula
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Parameters/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Parameters
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../Results/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Results
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="debug-and-refinement">Debug and Refinement<a class="headerlink" href="#debug-and-refinement" title="Permanent link">&para;</a></h1>
<h2 id="dependencies">Dependencies<a class="headerlink" href="#dependencies" title="Permanent link">&para;</a></h2>
<h3 id="r-environment">R environment<a class="headerlink" href="#r-environment" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>conda install -c conda-forge r-data.table
<span class="linenos" data-linenos="2 "></span>conda install -c conda-forge r-e1071
<span class="linenos" data-linenos="3 "></span>conda install -c conda-forge r-ggplot2
<span class="linenos" data-linenos="4 "></span>conda install -c conda-forge r-r.utils
</code></pre></div>
<h3 id="other-tools">Other tools<a class="headerlink" href="#other-tools" title="Permanent link">&para;</a></h3>
<p>Authors put the binary compatibility tools including <code>samtools, bcftools, beagle</code> in the app folder. If you want to specify tools by yourself, there are several modifications to do.<br />
I encountered some errors while running bcftools provided by the author and decided to use module in HPC, so take bcftools as an example here.<br />
Line <code>243,423</code> in <code>Monopogen.py</code>; Line <code>210</code> in <code>somatic.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>bcftools = os.path.abspath(args.app_path) + &quot;/bcftools&quot;
</code></pre></div>
<p>Revised to:</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>bcftools = &quot;bcftools&quot; # or path to your bcftools
</code></pre></div>
<h2 id="sex-chromosomes-is-not-considered-by-default">Sex chromosomes is not considered by default<a class="headerlink" href="#sex-chromosomes-is-not-considered-by-default" title="Permanent link">&para;</a></h2>
<p>If you consider the X chromosome, you need to modify the following code.</p>
<h4 id="the-way-to-specify-imputation-panel">The way to specify imputation panel<a class="headerlink" href="#the-way-to-specify-imputation-panel" title="Permanent link">&para;</a></h4>
<p>The source code has hardcoded the imputation panel.<br />
Line <code>76</code> in <code>Monopogen.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">imputation_vcf</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">imputation_panel</span> <span class="o">+</span> <span class="s2">&quot;CCDG_14151_B01_GRM_WGS_2020-08-05_&quot;</span> <span class="o">+</span> <span class="n">record</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;.filtered.shapeit2-duohmm-phased.vcf.gz” </span>
</code></pre></div>
<p>However, the naming convention for the X chromosome's imputation panel in 1KG differs from that of the autosomes. It has been modified to be specified externally.</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">imputation_vcf</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">imputation_panel</span>
</code></pre></div>
<p>After that, the code for judging the type of <code>args.imputation_panel</code> also needs to be modified.<br />
Line <code>92</code> in <code>germline.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">imputation_panel</span><span class="p">)</span>
</code></pre></div>
<p>Revised to:</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">imputation_panel</span><span class="p">)</span>
</code></pre></div>
<h4 id="add-x-chromosome-in-multiple-loops">Add X chromosome in multiple loops<a class="headerlink" href="#add-x-chromosome-in-multiple-loops" title="Permanent link">&para;</a></h4>
<p>Line <code>111,306,318</code> in <code>Monopogen.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="k">for</span> <span class="nb">chr</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">):</span>
<span class="linenos" data-linenos="2 "></span>    <span class="o">...</span>
</code></pre></div>
<p>Revised to:</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">chromosomes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">))</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">]</span>
<span class="linenos" data-linenos="2 "></span><span class="k">for</span> <span class="nb">chr</span> <span class="ow">in</span> <span class="n">chromosomes</span><span class="p">:</span>
<span class="linenos" data-linenos="3 "></span>    <span class="o">...</span>
</code></pre></div>
<h4 id="modify-source-file">Modify source file<a class="headerlink" href="#modify-source-file" title="Permanent link">&para;</a></h4>
<p>The resource file <code>Monopogen\resource\GRCh38.region.10MB.lst</code> and <code>Monopogen\resource\GRCh38.region.50MB.lst</code> does not contain the X chromosome also results in the X chromosome not being considered. <br />
Add lines below to <code>GRCh38.region.10MB.lst</code> for example.</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span>chrX,2,10000001
<span class="linenos" data-linenos=" 2 "></span>chrX,10000002,20000001
<span class="linenos" data-linenos=" 3 "></span>chrX,20000002,30000001
<span class="linenos" data-linenos=" 4 "></span>chrX,30000002,40000001
<span class="linenos" data-linenos=" 5 "></span>chrX,40000002,50000001
<span class="linenos" data-linenos=" 6 "></span>chrX,50000002,60000001
<span class="linenos" data-linenos=" 7 "></span>chrX,60000002,70000001
<span class="linenos" data-linenos=" 8 "></span>chrX,70000002,80000001
<span class="linenos" data-linenos=" 9 "></span>chrX,80000002,90000001
<span class="linenos" data-linenos="10 "></span>chrX,90000002,100000001
<span class="linenos" data-linenos="11 "></span>chrX,100000002,110000001
<span class="linenos" data-linenos="12 "></span>chrX,110000002,120000001
<span class="linenos" data-linenos="13 "></span>chrX,120000002,130000001
<span class="linenos" data-linenos="14 "></span>chrX,130000002,140000001
<span class="linenos" data-linenos="15 "></span>chrX,140000002,150000001
<span class="linenos" data-linenos="16 "></span>chrX,150000002,156040895
</code></pre></div>
<h2 id="rewrite-code-spliting-bam-by-cell-barcodes">Rewrite code spliting bam by cell barcodes<a class="headerlink" href="#rewrite-code-spliting-bam-by-cell-barcodes" title="Permanent link">&para;</a></h2>
<p>The original code, when splitting the original BAM file by CB, iterates through the BAM file as many times as there are different CBs, which is very inefficient.<br />
Therefore, the code has been rewritten to accomplish the task with just a single iteration through the BAM file.  </p>
<p>Substitute Line <code>160-196</code> (<code>def bamSplit</code>) in <code>somatic.py</code> by lines below.
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="k">def</span> <span class="nf">splitBam</span><span class="p">(</span><span class="n">cell_barcodes</span><span class="p">,</span> <span class="n">out_dir</span><span class="p">,</span> <span class="n">app_path</span><span class="p">):</span>
<span class="linenos" data-linenos=" 2 "></span>    <span class="n">samtools</span> <span class="o">=</span> <span class="n">app_path</span> <span class="o">+</span> <span class="s2">&quot;/samtools&quot;</span>
<span class="linenos" data-linenos=" 3 "></span>    <span class="n">output_bam</span> <span class="o">=</span>  <span class="n">out_dir</span> <span class="o">+</span> <span class="s2">&quot;/Bam/merge.filter.targeted.bam&quot;</span> 
<span class="linenos" data-linenos=" 4 "></span>    <span class="n">infile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">output_bam</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 5 "></span>    <span class="n">header</span> <span class="o">=</span> <span class="n">infile</span><span class="o">.</span><span class="n">header</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="linenos" data-linenos=" 6 "></span>
<span class="linenos" data-linenos=" 7 "></span>    <span class="n">avail_read</span> <span class="o">=</span> <span class="mi">0</span>
<span class="linenos" data-linenos=" 8 "></span>    <span class="c1"># create dictionary</span>
<span class="linenos" data-linenos=" 9 "></span>    <span class="n">outfiles</span> <span class="o">=</span> <span class="p">{}</span>
<span class="linenos" data-linenos="10 "></span>    <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cell_barcodes</span><span class="p">:</span>
<span class="linenos" data-linenos="11 "></span>        <span class="n">tp</span> <span class="o">=</span> <span class="n">infile</span><span class="o">.</span><span class="n">header</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
<span class="linenos" data-linenos="12 "></span>
<span class="linenos" data-linenos="13 "></span>        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tp</span><span class="p">[</span><span class="s1">&#39;RG&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
<span class="linenos" data-linenos="14 "></span>            <span class="n">tp</span><span class="p">[</span><span class="s1">&#39;RG&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">tp</span><span class="p">[</span><span class="s1">&#39;RG&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span>
<span class="linenos" data-linenos="15 "></span>        <span class="n">tp</span><span class="p">[</span><span class="s1">&#39;RG&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;SM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell</span>
<span class="linenos" data-linenos="16 "></span>        <span class="n">tp</span><span class="p">[</span><span class="s1">&#39;RG&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell</span>
<span class="linenos" data-linenos="17 "></span>
<span class="linenos" data-linenos="18 "></span>        <span class="n">outfile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span> <span class="s2">&quot;Bam/split_bam&quot;</span><span class="p">,</span> <span class="n">cell</span> <span class="o">+</span> <span class="s2">&quot;.bam&quot;</span><span class="p">),</span> <span class="s2">&quot;wb&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">tp</span><span class="p">)</span>
<span class="linenos" data-linenos="19 "></span>        <span class="n">outfiles</span><span class="p">[</span><span class="n">cell</span><span class="p">]</span> <span class="o">=</span> <span class="n">outfile</span>
<span class="linenos" data-linenos="20 "></span>
<span class="linenos" data-linenos="21 "></span>    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
<span class="linenos" data-linenos="22 "></span>        <span class="n">cell_barcode</span> <span class="o">=</span> <span class="n">robust_get_tag</span><span class="p">(</span><span class="n">read</span><span class="p">,</span> <span class="s2">&quot;CB&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="23 "></span>        <span class="k">if</span> <span class="n">cell_barcode</span> <span class="ow">in</span> <span class="n">outfiles</span><span class="p">:</span>
<span class="linenos" data-linenos="24 "></span>            <span class="n">outfiles</span><span class="p">[</span><span class="n">cell_barcode</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>
<span class="linenos" data-linenos="25 "></span>            <span class="n">avail_read</span> <span class="o">=</span> <span class="n">avail_read</span> <span class="o">+</span> <span class="mi">1</span>
<span class="linenos" data-linenos="26 "></span>
<span class="linenos" data-linenos="27 "></span>    <span class="k">for</span> <span class="n">outfile</span> <span class="ow">in</span> <span class="n">outfiles</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
<span class="linenos" data-linenos="28 "></span>        <span class="n">outfile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="linenos" data-linenos="29 "></span>    <span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="linenos" data-linenos="30 "></span>
<span class="linenos" data-linenos="31 "></span>    <span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cell_barcodes</span><span class="p">:</span>
<span class="linenos" data-linenos="32 "></span>        <span class="n">cmd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">samtools</span><span class="si">}</span><span class="s2"> index </span><span class="si">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">out_dir</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;Bam/split_bam&#39;</span><span class="p">,</span><span class="w"> </span><span class="n">cell</span><span class="p">)</span><span class="si">}</span><span class="s2">.bam&quot;</span>
<span class="linenos" data-linenos="33 "></span>        <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
<span class="linenos" data-linenos="34 "></span>
<span class="linenos" data-linenos="35 "></span>    <span class="k">return</span><span class="p">(</span><span class="n">avail_read</span><span class="p">)</span>
</code></pre></div></p>
<p>Then substitute Line <code>188-200</code> in <code>Monopogen.py</code> by lines below.</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">result</span> <span class="o">=</span> <span class="n">splitBam</span><span class="p">(</span><span class="n">cell_lst</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">app_path</span><span class="p">)</span>
<span class="linenos" data-linenos="2 "></span><span class="k">if</span> <span class="n">result</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos="3 "></span>    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;No reads detected for cells in &quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">barcode</span> <span class="o">+</span> <span class="s2">&quot;. Please check 1) the input cell barcode file is matched with bam file; 2) the cell barcode name has the same format shown in bam file. For example XX-1!&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="4 "></span>    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed! See instructions above.&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="5 "></span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>
<details class="question">
<summary>Line <code>188-200</code> in <code>Monopogen.py</code></summary>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="n">joblst</span> <span class="o">=</span> <span class="p">[]</span>
<span class="linenos" data-linenos=" 2 "></span>
<span class="linenos" data-linenos=" 3 "></span><span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cell_lst</span><span class="p">:</span>
<span class="linenos" data-linenos=" 4 "></span>        <span class="n">para</span> <span class="o">=</span> <span class="s2">&quot;merge&quot;</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">cell</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">app_path</span>
<span class="linenos" data-linenos=" 5 "></span>        <span class="n">joblst</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">para</span><span class="p">)</span>
<span class="linenos" data-linenos=" 6 "></span>
<span class="linenos" data-linenos=" 7 "></span><span class="k">with</span> <span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">nthreads</span><span class="p">)</span> <span class="k">as</span> <span class="n">pool</span><span class="p">:</span>
<span class="linenos" data-linenos=" 8 "></span>    <span class="n">result</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">bamSplit</span><span class="p">,</span> <span class="n">joblst</span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span>
<span class="linenos" data-linenos="10 "></span><span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">result</span><span class="p">)</span><span class="o">==</span><span class="mi">0</span><span class="p">:</span>
<span class="linenos" data-linenos="11 "></span>    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;No reads detected for cells in &quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">barcode</span> <span class="o">+</span> <span class="s2">&quot;. Please check 1) the input cell barcode file is matched with bam file; 2) the cell barcode name has the same format shown in bam file. For example XX-1!&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="12 "></span>    <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Failed! See instructions above.&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span>    <span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</code></pre></div>
</details>
<h2 id="other-bugs">Other bugs<a class="headerlink" href="#other-bugs" title="Permanent link">&para;</a></h2>
<h3 id="havent-generate-index-of-vcf-file">Haven't generate index of vcf file.<a class="headerlink" href="#havent-generate-index-of-vcf-file" title="Permanent link">&para;</a></h3>
<p>Add lines above after Line 85 in <code>Monopogen.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">cmd6</span> <span class="o">=</span> <span class="n">tabix</span> <span class="o">+</span> <span class="s2">&quot; -p vcf &quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;/germline/&quot;</span> <span class="o">+</span>  <span class="n">jobid</span> <span class="o">+</span> <span class="s2">&quot;.gl.vcf.gz; &quot;</span> <span class="o">+</span> <span class="n">tabix</span> <span class="o">+</span> <span class="s2">&quot; -p vcf &quot;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;/germline/&quot;</span> <span class="o">+</span> <span class="n">jobid</span> <span class="o">+</span> <span class="s2">&quot;.phased.vcf.gz&quot;</span>
<span class="linenos" data-linenos="2 "></span><span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">cmd6</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</code></pre></div>
<h3 id="separator-is-not-specified-in-read_csv">Separator is not specified in <code>read_csv</code>.<a class="headerlink" href="#separator-is-not-specified-in-read_csv" title="Permanent link">&para;</a></h3>
<p>The cell barcode file includes two column: 1.cell barcode; 2.number of reads detected in each cell (separated by tab).<br />
Line <code>179-180</code> in <code>Monopogen.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">cell_clst</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">barcode</span><span class="p">)</span>   
<span class="linenos" data-linenos="2 "></span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">cell_clst</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cell&#39;</span><span class="p">,</span><span class="s1">&#39;id&#39;</span><span class="p">])</span>
</code></pre></div>
<p>Revised to:</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>df = pd.read_csv(args.barcode,sep=&quot;\t&quot;,header=0)
</code></pre></div>
<h3 id="indent-error">Indent error<a class="headerlink" href="#indent-error" title="Permanent link">&para;</a></h3>
<p>Line <code>190</code> in <code>Monopogen.py</code></p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="k">for</span> <span class="n">cell</span> <span class="ow">in</span> <span class="n">cell_lst</span><span class="p">:</span>
<span class="linenos" data-linenos="2 "></span>                <span class="n">para</span> <span class="o">=</span> <span class="s2">&quot;merge&quot;</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cell</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">out</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">app_path</span>
<span class="linenos" data-linenos="3 "></span>                <span class="n">joblst</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">para</span><span class="p">)</span>
</code></pre></div>
<h3 id="release-the-limit-on-the-number-of-files-that-can-be-opened-simultaneously">Release the limit on the number of files that can be opened simultaneously<a class="headerlink" href="#release-the-limit-on-the-number-of-files-that-can-be-opened-simultaneously" title="Permanent link">&para;</a></h3>
<p>During the joint calling process, the BAM file of each cell is opened. When the number of cells is large, this can exceed the limit of the number of files that can be opened by the Linux system simultaneously. The default value is set to 1048. When the number of your cells exceeds this value, it needs to be adjusted.</p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span>ulimit -n 2048 # According to your cell barcode count.
</code></pre></div>
<h2 id="issues-in-ldrefinementr">Issues in LDrefinement.R<a class="headerlink" href="#issues-in-ldrefinementr" title="Permanent link">&para;</a></h2>
<h3 id="issues-in-define-snv-pairs">Issues in define SNV pairs<a class="headerlink" href="#issues-in-define-snv-pairs" title="Permanent link">&para;</a></h3>
<p>When germline SNV pairs found within the current cell do not surround any de novo SNVs with reads, the process still proceeds to calculations for somatic SNVs. To improve efficiency, an additional layer of judgment is added to the calculations for de novo SNVs.  </p>
<p>Line <code>113-129</code> and Line <code>162-177</code> in <code>LDrefinement.R</code>  </p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">somaticIndex_i</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">){</span>
<span class="linenos" data-linenos="2 "></span><span class="w">    </span><span class="c1"># Line 113-129 or Line 162-177</span>
<span class="linenos" data-linenos="3 "></span><span class="p">}</span>
</code></pre></div>
<p>In the trioloci model, a list of germline SNVs with reads, sorted by coordinates, is first extracted. Then, the nearest downstream germline SNV to the de novo SNV is selected. It is directly assumed that another SNV in the list with a larger coordinate than the first one forms a triolocus with it. However, there is no check to confirm whether the coordinates of this SNV are indeed larger than those of the de novo SNV, which may lead to issues. Add a conditional statement to make process more robust.  </p>
<p>Line <code>169</code> in <code>LDrefinement.R</code><br />
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">if</span><span class="p">(</span><span class="n">mat</span><span class="p">[</span><span class="n">lower1</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="o">==</span><span class="n">mat</span><span class="p">[</span><span class="n">upper1</span><span class="p">,</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&amp;&amp;</span><span class="w"> </span><span class="n">dis</span><span class="p">[</span><span class="n">upper1</span><span class="p">]</span><span class="o">&gt;</span><span class="n">dis</span><span class="p">[</span><span class="n">j</span><span class="p">]){</span><span class="kc">...</span><span class="p">}</span>
</code></pre></div></p>
<h3 id="issue-in-the-funtion-weighthedp">Issue in the funtion <code>weighthedP</code><a class="headerlink" href="#issue-in-the-funtion-weighthedp" title="Permanent link">&para;</a></h3>
<p>The function <code>weighthedP</code> is designed to calculate the LD-refinement score of de novo SNVs based on the LD-refinement score of germline SNV pairs. However, it's nonsensical that the bin size for physical distance used here is different from the one used when calculating the LD-refinement score for germline SNV pairs. This needs to be modified.  </p>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="c1"># binsize in calculating germline LD-refinement score (Line 222 in LDrefinement.R)</span>
<span class="linenos" data-linenos="2 "></span><span class="n">binsize</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">200</span><span class="p">,</span><span class="w"> </span><span class="m">300</span><span class="p">,</span><span class="w"> </span><span class="m">400</span><span class="p">,</span><span class="w"> </span><span class="m">500</span><span class="p">,</span><span class="w"> </span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="m">2500</span><span class="p">,</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="m">7500</span><span class="p">,</span><span class="w"> </span><span class="m">10000</span><span class="p">,</span><span class="w"> </span><span class="m">20000</span><span class="p">,</span><span class="w"> </span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="m">500000</span><span class="p">,</span><span class="w"> </span><span class="m">1000000000000000000</span><span class="p">)</span>
<span class="linenos" data-linenos="3 "></span><span class="c1"># binsize in calculating somatic LD-refinement score (Line 204 in LDrefinement.R)</span>
<span class="linenos" data-linenos="4 "></span><span class="n">binsize</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="m">100</span><span class="p">,</span><span class="w"> </span><span class="m">500</span><span class="p">,</span><span class="w"> </span><span class="m">1000</span><span class="p">,</span><span class="w"> </span><span class="m">2500</span><span class="p">,</span><span class="w"> </span><span class="m">5000</span><span class="p">,</span><span class="w"> </span><span class="m">7500</span><span class="p">,</span><span class="w"> </span><span class="m">10000</span><span class="p">,</span><span class="w"> </span><span class="m">20000</span><span class="p">,</span><span class="w"> </span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="m">500000</span><span class="p">,</span><span class="w"> </span><span class="m">1000000000000000000</span><span class="p">)</span>
</code></pre></div>
<p>Moreover, when indexing the germline LD-refinement score corresponding to the distance of the loci pair, the index is incorrect and should be <code>i-1</code>.  </p>
<p>Line <code>213</code> in <code>LDrefinement.R</code><br />
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">rd_w</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">rd_w</span><span class="p">,</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">table</span><span class="o">$</span><span class="n">Prob</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">^</span><span class="m">2</span><span class="p">)</span>
</code></pre></div>
Revised to
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="n">rd_w</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">rd_w</span><span class="p">,</span><span class="w"> </span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">table</span><span class="o">$</span><span class="n">Prob</span><span class="p">[</span><span class="n">i</span><span class="m">-1</span><span class="p">])</span><span class="o">^</span><span class="m">2</span><span class="p">)</span>
</code></pre></div></p>
<h3 id="issue-in-adjusting-ld-refinement-score-based-on-their-phasing-information">Issue in adjusting LD refinement score based on their phasing information<a class="headerlink" href="#issue-in-adjusting-ld-refinement-score-based-on-their-phasing-information" title="Permanent link">&para;</a></h3>
<p>When refining the final score for each site, if one of the two-loci and trio-loci LD-refinement scores is missing, the final score is assigned as NA, which is not quite reasonable. It has been modified to be assigned as the existing score among the two.<br />
Line <code>346-353</code> in <code>LDrefinement.R</code>  </p>
<p><div class="highlight"><pre><span></span><code><span class="linenos" data-linenos="1 "></span><span class="nf">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="nf">nrow</span><span class="p">(</span><span class="n">dt</span><span class="p">),</span><span class="m">1</span><span class="p">)){</span>
<span class="linenos" data-linenos="2 "></span><span class="w">   </span><span class="nf">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">])){</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">&lt;-</span><span class="m">10</span><span class="p">}</span>
<span class="linenos" data-linenos="3 "></span><span class="w">   </span><span class="nf">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">])){</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">&lt;-</span><span class="m">10</span><span class="p">}</span>
<span class="linenos" data-linenos="4 "></span><span class="w">   </span><span class="n">dt</span><span class="o">$</span><span class="n">p_LDrefine</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">+</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">*</span><span class="m">0.5</span>
<span class="linenos" data-linenos="5 "></span><span class="w">   </span><span class="nf">if</span><span class="p">(</span><span class="n">dt</span><span class="o">$</span><span class="n">p_LDrefine</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">&gt;</span><span class="m">1</span><span class="p">){</span>
<span class="linenos" data-linenos="6 "></span><span class="w">    </span><span class="n">dt</span><span class="o">$</span><span class="n">p_LDrefine</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="kc">NA</span>
<span class="linenos" data-linenos="7 "></span><span class="w">   </span><span class="p">}</span>
<span class="linenos" data-linenos="8 "></span><span class="p">}</span>
</code></pre></div>
Revised to
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="nf">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="nf">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="nf">nrow</span><span class="p">(</span><span class="n">dt</span><span class="p">),</span><span class="m">1</span><span class="p">)){</span>
<span class="linenos" data-linenos=" 2 "></span><span class="w">   </span><span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="w"> </span><span class="o">&amp;</span><span class="w">  </span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">])){</span>
<span class="linenos" data-linenos=" 3 "></span><span class="w">     </span><span class="n">dt</span><span class="o">$</span><span class="n">p_LDrefine</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">/</span><span class="m">2</span>
<span class="linenos" data-linenos=" 4 "></span><span class="w">   </span><span class="p">}</span>
<span class="linenos" data-linenos=" 5 "></span><span class="w">   </span><span class="nf">if</span><span class="p">(</span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="w"> </span><span class="o">&amp;</span><span class="w">  </span><span class="nf">is.na</span><span class="p">(</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">])){</span>
<span class="linenos" data-linenos=" 6 "></span><span class="w">     </span><span class="n">dt</span><span class="o">$</span><span class="n">p_LDrefine</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="linenos" data-linenos=" 7 "></span><span class="w">   </span><span class="p">}</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">   </span><span class="nf">if</span><span class="p">(</span><span class="nf">is.na</span><span class="p">(</span><span class="n">a</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="w"> </span><span class="o">&amp;</span><span class="w">  </span><span class="o">!</span><span class="nf">is.na</span><span class="p">(</span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">])){</span>
<span class="linenos" data-linenos=" 9 "></span><span class="w">     </span><span class="n">dt</span><span class="o">$</span><span class="n">p_LDrefine</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">b</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
<span class="linenos" data-linenos="10 "></span><span class="w">   </span><span class="p">}</span>
<span class="linenos" data-linenos="11 "></span><span class="p">}</span>
</code></pre></div></p>
<h2 id="paint-germline-ld-refinement-score-plot-of-all-chromosomes-in-a-picture">Paint germline LD refinement score plot of all chromosomes in a picture<a class="headerlink" href="#paint-germline-ld-refinement-score-plot-of-all-chromosomes-in-a-picture" title="Permanent link">&para;</a></h2>
<p>Author didn't provide the code painting germline LD refinement score plot of all chromosomes in a picture like below. You can use the script below directly.  </p>
<figure>
<p><img alt="Image title" src="../monopogen_image/DNA_twoloci_allchr_paper.png" width="50%" />
  </p>
<figcaption>Twoloci model from paper</figcaption>
</figure>
<div class="highlight"><pre><span></span><code><span class="linenos" data-linenos=" 1 "></span><span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="linenos" data-linenos=" 2 "></span>
<span class="linenos" data-linenos=" 3 "></span><span class="n">dir</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s">&quot;path/to/monopogen/somatic/&quot;</span>
<span class="linenos" data-linenos=" 4 "></span><span class="n">chr_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">22</span><span class="p">,</span><span class="s">&#39;X&#39;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 5 "></span><span class="n">all_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">data.frame</span><span class="p">()</span>
<span class="linenos" data-linenos=" 6 "></span>
<span class="linenos" data-linenos=" 7 "></span><span class="nf">for </span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="n">in</span><span class="w"> </span><span class="n">chr_list</span><span class="p">)</span><span class="w"> </span><span class="p">{</span>
<span class="linenos" data-linenos=" 8 "></span><span class="w">  </span><span class="n">file_name</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;.germlineTrioLoci_model.csv&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos=" 9 "></span><span class="w">  </span><span class="n">data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">read.csv</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span><span class="w"> </span><span class="n">header</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">TRUE</span><span class="p">)</span>
<span class="linenos" data-linenos="10 "></span><span class="w">  </span><span class="n">data</span><span class="o">$</span><span class="n">bin</span><span class="p">[</span><span class="n">data</span><span class="o">$</span><span class="n">bin</span><span class="o">&gt;</span><span class="m">5</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="m">5</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">50</span><span class="o">*</span><span class="m">10</span><span class="o">^</span><span class="m">5</span>
<span class="linenos" data-linenos="11 "></span><span class="w">  </span><span class="n">data</span><span class="o">$</span><span class="n">Chromosome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.factor</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="s">&quot;Chr&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">i</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">))</span>
<span class="linenos" data-linenos="12 "></span><span class="w">  </span><span class="n">all_data</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">rbind</span><span class="p">(</span><span class="n">all_data</span><span class="p">,</span><span class="w"> </span><span class="n">data</span><span class="p">)</span>
<span class="linenos" data-linenos="13 "></span><span class="p">}</span>
<span class="linenos" data-linenos="14 "></span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">ggplot</span><span class="p">(</span><span class="n">all_data</span><span class="p">,</span><span class="w"> </span><span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">bin</span><span class="p">,</span><span class="w"> </span><span class="n">y</span><span class="o">=</span><span class="n">Prob</span><span class="p">))</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="15 "></span><span class="w">  </span><span class="nf">geom_point</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;cadetblue&quot;</span><span class="p">,</span><span class="n">size</span><span class="o">=</span><span class="m">6</span><span class="p">)</span><span class="o">+</span>
<span class="linenos" data-linenos="16 "></span><span class="w">  </span><span class="nf">geom_smooth</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="s">&quot;darksalmon&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">scale_x_continuous</span><span class="p">(</span><span class="n">trans</span><span class="o">=</span><span class="s">&#39;log10&#39;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="17 "></span><span class="w">  </span><span class="nf">theme_bw</span><span class="p">()</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span><span class="n">panel.border</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span><span class="w"> </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="linenos" data-linenos="18 "></span><span class="w">                     </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span><span class="w"> </span><span class="n">axis.line</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_line</span><span class="p">(</span><span class="n">colour</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">xlab</span><span class="p">(</span><span class="s">&quot;Physical distance (bp)&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="19 "></span><span class="w">  </span><span class="nf">ylab</span><span class="p">(</span><span class="s">&quot;LD refinement score&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="20 "></span><span class="w">  </span><span class="nf">ylim</span><span class="p">(</span><span class="m">0</span><span class="p">,</span><span class="m">0.65</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="21 "></span><span class="w">  </span><span class="nf">facet_wrap</span><span class="p">(</span><span class="o">~</span><span class="w"> </span><span class="n">Chromosome</span><span class="p">,</span><span class="w"> </span><span class="n">ncol</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">6</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="22 "></span><span class="w">  </span><span class="nf">labs</span><span class="p">(</span><span class="n">title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;Three-locus LD refinement score on germline SNVs&quot;</span><span class="p">)</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="23 "></span><span class="w">  </span><span class="nf">theme_minimal</span><span class="p">()</span><span class="w"> </span><span class="o">+</span>
<span class="linenos" data-linenos="24 "></span><span class="w">  </span><span class="nf">theme</span><span class="p">(</span>
<span class="linenos" data-linenos="25 "></span><span class="w">    </span><span class="n">strip.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="linenos" data-linenos="26 "></span><span class="w">    </span><span class="n">strip.background</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_rect</span><span class="p">(</span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;lightgray&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">),</span>
<span class="linenos" data-linenos="27 "></span><span class="w">    </span><span class="n">plot.title</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">0.5</span><span class="p">,</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">face</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;bold&quot;</span><span class="p">),</span>
<span class="linenos" data-linenos="28 "></span><span class="w">    </span><span class="n">axis.text.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="linenos" data-linenos="29 "></span><span class="w">    </span><span class="n">axis.text.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="linenos" data-linenos="30 "></span><span class="w">    </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="linenos" data-linenos="31 "></span><span class="w">    </span><span class="n">axis.title.x</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">18</span><span class="p">),</span>
<span class="linenos" data-linenos="32 "></span><span class="w">    </span><span class="n">axis.title.y</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">size</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">18</span><span class="p">),</span>
<span class="linenos" data-linenos="33 "></span><span class="w">    </span><span class="n">panel.grid.major</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="linenos" data-linenos="34 "></span><span class="w">    </span><span class="n">panel.grid.minor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_blank</span><span class="p">(),</span>
<span class="linenos" data-linenos="35 "></span><span class="w">    </span><span class="n">panel.spacing</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">unit</span><span class="p">(</span><span class="m">0.5</span><span class="p">,</span><span class="w"> </span><span class="s">&quot;lines&quot;</span><span class="p">)</span>
<span class="linenos" data-linenos="36 "></span><span class="w">  </span><span class="p">)</span>
<span class="linenos" data-linenos="37 "></span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">p</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="nf">theme</span><span class="p">(</span>
<span class="linenos" data-linenos="38 "></span><span class="w">  </span><span class="n">axis.ticks</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_line</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">),</span>
<span class="linenos" data-linenos="39 "></span><span class="w">  </span><span class="n">axis.text.y.left</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">18</span><span class="p">),</span>
<span class="linenos" data-linenos="40 "></span><span class="w">  </span><span class="n">axis.text.x.bottom</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">element_text</span><span class="p">(</span><span class="n">angle</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">90</span><span class="p">,</span><span class="w"> </span><span class="n">hjust</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;black&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">size</span><span class="o">=</span><span class="m">18</span><span class="p">)</span>
<span class="linenos" data-linenos="41 "></span><span class="p">)</span>
<span class="linenos" data-linenos="42 "></span><span class="nf">print</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="linenos" data-linenos="43 "></span><span class="nf">ggsave</span><span class="p">(</span><span class="nf">paste0</span><span class="p">(</span><span class="n">dir</span><span class="p">,</span><span class="s">&quot;LDrefinement_germline_trioloci_allchr.pdf&quot;</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">device</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s">&quot;pdf&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">15</span><span class="p">,</span><span class="w"> </span><span class="n">dpi</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">300</span><span class="p">)</span>
</code></pre></div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.sections", "toc.integrate", "navigation.top", "search.suggest", "search.highlight", "content.tabs.link", "content.code.annotation", "content.code.copy"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.8fd75fb4.min.js"></script>
      
        <script src="../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>